### PR TITLE
Pydantic v2 migration

### PR DESCRIPTION
**Pull request by mrwyattii**
Pydantic v2 has been out for some time now. We have been relying on using the v1 API available in v2 until now. This is a refresh of #3902 to bring proper v2 support to DeepSpeed.

Corresponding DeepSpeed-MII PR [here](https://github.com/microsoft/DeepSpeed-MII/pull/423).

@loadams

### CODE REVIEW

File: requirements/requirements.txt [hunk 1]
```diff
@@ -4,7 +4,7 @@ numpy
 packaging>=20.0
 psutil
 py-cpuinfo
-pydantic
+pydantic>=2.0.0
```
Comment (loadams) [inline requirements/requirements.txt:7]: Should we update the pydantic version requirement in requirements-readthedocs?

**Comment by loadams** [comment]
Do you want to do what we did on the previous pydantic update of updating the nv-mii test to use your branch rather than `main` so that it passes here?

File: .github/workflows/nv-a6000.yml [hunk 2]
```diff
@@ -47,6 +47,7 @@ jobs:
       - name: Install deepspeed
         run: |
           python -m pip install docutils==0.18.1 jinja2==3.0 urllib3==1.26.11 ninja
+          python -m pip install thinc spacy confection --upgrade
```
Comment (loadams) [inline .github/workflows/nv-a6000.yml:50]: Previous versions of these packages didn't support pydantic > 2, so this works around the issue for now, ideally updating them in the container image in another PR.

**Comment by mrwyattii** [comment]
CI with microsoft/DeepSpeed-MII#423 for [nv-a6000](https://github.com/microsoft/DeepSpeed/actions/runs/8118779187/job/22193652085) and [nv-mii](https://github.com/microsoft/DeepSpeed/actions/runs/8118827112/job/22193779268)

File: .github/workflows/nv-a6000.yml (hunk 2 above)
Comment (loadams) [reply .github/workflows/nv-a6000.yml:50]: @mrwyattii - is it worth adding a comment here for why these were upgraded?

File: .github/workflows/python.yml [hunk 3]
```diff
@@ -21,7 +21,7 @@ jobs:
   unit-tests:
     strategy:
       matrix:
-        pyVersion: ["3.6", "3.7", "3.8", "3.9", "3.10"]
+        pyVersion: ["3.7", "3.8", "3.9", "3.10"]
```
Comment (loadams) [inline .github/workflows/python.yml:24]: Is it worth adding 3.11 to this at some point?

File: deepspeed/runtime/config_utils.py [hunk 4]
```diff
@@ -54,67 +55,73 @@ def __init__(self, strict=False, **data):
...
         # Get information about the deprecated field
-        fields_set = pydantic_config.__fields_set__
-        dep_param = field.name
-        kwargs = field.field_info.extra
+        pydantic_config = self
+        fields_set = pydantic_config.model_fields_set
+        kwargs = pydantic_config.model_fields[dep_field].json_schema_extra
         new_param_fn = kwargs.get("new_param_fn", lambda x: x)
-        param_value = new_param_fn(getattr(pydantic_config, dep_param))
-        new_param = kwargs.get("new_param", "")
```
Comment (loadams) [inline deepspeed/runtime/config_utils.py:66]: Just curious, why couldn't it stay as new_param?

**Review by loadams — APPROVED** [review]

File: .github/workflows/nv-a6000.yml [hunk 5]
```diff
@@ -23,7 +23,7 @@ jobs:
   unit-tests:
     runs-on: [self-hosted, nvidia, a6000]
     container:
-      image: nvcr.io/nvidia/pytorch:23.03-py3
+      image: nvcr.io/nvidia/pytorch:23.09-py3
```
Comment (loadams) [inline .github/workflows/nv-a6000.yml:26]: Updating container to have pydantic 2+ support and updated packages of thinc, spacy, and confection that all were tied to specific pydantic versions.

File: .github/workflows/nv-a6000.yml (hunk 2 above)
Comment (loadams) [reply .github/workflows/nv-a6000.yml:51]: Added.

**Review by loadams — APPROVED** [review]

**Comment by FrsECM** [comment]
Using Albumentation, the last version has a strict dependancy regarding pydantic >2.0 what makes it incompatible with deepspeed.

https://github.com/albumentations-team/albumentations/releases/tag/1.4.4

https://github.com/albumentations-team/albumentations/blob/a22a319b2e7079389326e5b7830e71017701b1b9/setup.py#L11

It makes this PR even more releavant.

**Review by lekurile — APPROVED** [review]

File: .github/workflows/python.yml (hunk 3 above)
Comment (loadams) [reply .github/workflows/python.yml:24]: Yes, but moving this to another PR.

**Comment by loadams** [comment]
CI triggered using updates in MII

[nv-a6000](https://github.com/microsoft/DeepSpeed/actions/runs/9751396435)

**Comment by loadams** [comment]
Tested here

https://github.com/microsoft/DeepSpeed/actions/runs/10512334326
https://github.com/microsoft/DeepSpeed/actions/runs/10512339060

**Pull request merged**